src/otx/core/data/dataset/base.py

```diff
@@ -0,0 +1,80 @@
+"""Base class for OTXDataset."""
+from __future__ import annotations
+
+from abc import abstractmethod
+from collections.abc import Iterable
+from typing import TYPE_CHECKING, Callable, Generic, List, Optional, Union
+
+import numpy as np
+from torch.utils.data import Dataset
+
+from otx.core.data.entity.base import T_OTXDataEntity
+
+if TYPE_CHECKING:
+    from datumaro import DatasetSubset
+
+Transforms = Union[Callable, List[Callable]]
+
+
+class OTXDataset(Dataset, Generic[T_OTXDataEntity]):
```

Comment:
`OTXDataset` class appears to contain a significant amount of logic specific to the mmframework, which can make initial comprehension challenging. 
This complexity could hinder scalability in the future. It might be more beneficial to maintain `mmengie.BaseDataset` class, rather than developing a logic to accommodate various frameworks.

Reply:
Sorry, but I don't get your point. What you are saying is like abandoning the principle that there should be one data entity I proposed. As you know, I proved that we can reuse the MMDet's data transforms in this design. In addition, `OTXDataset` does not contain a significant amount of logic specific to `mm*`. `mm*`'s transforms are actually adapted here `src/otx/core/data/transform_libs/mmdet.py`. `OTXDataset` is just calling it.

If you are saying about leaving `Transforms` into single type such as `Transforms = Union[Callable, List[Callable]] => Transforms = List[Callable]]`. Yea, I'm open to it. We can do it at the next PR.

Reply:
Indeed, the transform-related code seems to be a significant part of this class.

If we are dealing with multiple different frameworks, it's not immediately clear whether we would need to write multiple transform methods. It would be more efficient to find a method to integrate these.

The use of Iterable in this context doesn't necessarily have to be specific to mmengine. An Iterable could simply be a list. It's not entirely clear why this case is applied specifically to mmengine.

```python
if isinstance(self.transforms, Iterable):
    return self._mmengine_transforms(entity)
```

Reply:
Yea, it's fair point that the list of callables is not specific one for mmengine. Do you agree that renaming it to something another such as `_mmengine_transforms() -> _iterable_transforms()` at this stage?

Reply:
Renamed in https://github.com/openvinotoolkit/training_extensions/pull/2670/commits/dd5c72741eebd3f020383499f5edad8f3283f220